src/input.rs

```diff
@@ -578,11 +706,11 @@ impl<F: Clone, I: ?Sized> Clone for Filter<F, I> {
     }
 }
 
-pub fn filter<F: Fn(&I::Token) -> bool, I: Input + ?Sized>(filter: F) -> Filter<F, I> {
+pub fn filter<F: Fn(&I::Token) -> bool, I: Input + ?Sized, E: Error<I::Token>>(filter: F) -> Filter<F, I, E> {
```

Comment:
Could also add the state param, if you want. I don't use it so I'm not sure if it's relevant here.